### PR TITLE
fix: allow pricing update

### LINKED ISSUE #87: awsmp public-offer update/update-instance-type --allow-price-change not respected
**Describe the bug**
The --allow-price-change flag is intended to allow the following use cases:
1. enable migrating from free to paid
2. update pricing of existing instance types
3. add instance types to the listing

The logic currently prevents updating a listing unless its in the `Draft` stage. The only stage that is not allowed is `Restricted`

**To Reproduce**
1. run `awsmp public-offer [update | update-instance-type] --allow-price-change with a config file with diffs

**Expected behavior**
`--alow-price-change` correctly allows changes

### PR DESCRIPTION
**Pull request by rpocase**
fixes #87 

Summary of changes:
* Update allow price change to be a proper flag (previously always prompted)
* Generally respect --no-allow-price-change
* Ensure non-restricted listings can have instance values changed
* Update test cases to better match intention 
* Drop some code that did nothing (e.g. checking for `None` changeset against function that always return list)

# Exploratory testing

## Passing
* remove instance-type for paid listing (--allow-price-change)
* update pricing of instance-type for paid listing (--allow-price-change)
* ensure --no-allow-price-change rejects changes
   * adding new instance types
   * removing instance types
   * change hourly/annual prices for paid listing
   * prevent free to paid
* ensure prompt for price change is still available
* convert free product to paid product (while product is in staging)

## Failing
* ensure --no-allow-price-change rejects changes
   * prevent paid listing to free listing (this is as designed right now)

## Not tested
* remove product for free listing. I believe #89 is a real problem, but don't want to test against a limited/live listing for the scope of this PR

### CODE REVIEW

**Comment by rpocase** [comment]
i need to revisit some test cases after reverting the paid -> zero overrides. i think there is value in the tests but they need to be reworked a bit. i'll take care of that monday.

File: awsmp/_driver.py [hunk 1]
```diff
@@ -60,21 +60,14 @@ def update_instance_types(
...
 
-        if hourly_diff or annual_diff:
-            if not price_change_allowed:
-                logger.error(
-                    "There are pricing changes but changing price flag is not set. Please check the pricing files or set the price flag.\n"
-                    "Price change details:\n"
-                    f"Hourly: {hourly_diff}\nAnnual: {annual_diff}\n"
-                )
-                return None
-
```
Comment (rpocase) [inline awsmp/_driver.py:77]: after changes this had zero coverage. as far as i can tell, we can't hit this path anymore due to `_get_pricing_diff` raising in all matching cases

File: awsmp/_driver.py [hunk 2]
```diff
@@ -323,19 +319,37 @@ def _get_pricing_diff(product_id: str, changeset: List[ChangeSetType]) -> Tuple[
...
+        logger.error(error_msg)
+        raise AmiPriceChangeError(error_msg)
+    elif instance_configuration_changed and not allow_price_update:
+        error_message = f"""There are pricing changes in either hourly or annual prices.
+        Please check the pricing files or allow price change.
+        Price change details:\n
+        Local pricing updates: {local_annual}\Existing pricing in local: {existing_annual}\n"
+        """
+        logger.error(error_message)
+        raise AmiPriceChangeError(error_message)
```
Comment (rthill91) [inline awsmp/_driver.py:352]: If I'm reading correctly instance_configuration_changed will flag on the the addition or removal of annual prices (e.g. switching from hourly+annual to just hourly or vice versa). Those changes must be done via support ticket.

Breaking those cases out into a separate elif and raising a different error with similar messaging to the old AmiPriceChangeError should be sufficient (`AmiPriceModelChangeError` perhaps?)

**Review by rthill91 — CHANGES_REQUESTED** [review]

File: awsmp/_driver.py (hunk 2 above)
Comment (rpocase) [reply awsmp/_driver.py]: :done:

**Review by rthill91 — APPROVED** [review]

File: awsmp/_driver.py [hunk 3]
```diff
@@ -323,19 +319,37 @@ def _get_pricing_diff(product_id: str, changeset: List[ChangeSetType]) -> Tuple[
     diffs_hourly = _build_pricing_diff(existing_hourly, local_hourly)
     diffs_annual = _build_pricing_diff(existing_annual, local_annual)
 
-    if existing_listing_status != "Draft":
```
Comment (JessicaJang) [inline awsmp/_driver.py:326]: you can change the pricing type when the item is in draft status, as described in the comment

Comment (rpocase) [reply awsmp/_driver.py:326]: that should be accounted for in the new logic (see l#333)
